### ISSUE TITLE
Bazel: Build issue with npm dependency acorn

### ISSUE THREAD

**Issue by bkotsopoulossc** [emscripten-core/emsdk#999]
I am using the bazel SDK on version 3.1.3, and when I try upgrading to 3.1.7, I get the following build error, when compiling a C++ file:

```
Compiling <some C/C++ file> failed: missing input file 'external/emscripten_bin_linux/emscripten/node_modules/.bin/acorn', owner: '@emscripten_bin_linux//:emscripten/node_modules/.bin/acorn': /run/jenkins/bazel/userroot/c86299e60866b3fc0677d012d7096948/external/emscripten_bin_linux/emscripten/node_modules/.bin/acorn (No such file or directory)
```

I'm not sure why acorn is a dependency in the C++ compilation step, since as I understand it is a JS parser.

In this diff https://github.com/emscripten-core/emsdk/compare/3.1.3..3.1.7, it looks like the node version was updated, but I don't see any other major changes here that would related to this.

**Comment by bkotsopoulossc** [comment]
@mattsoulanille have you seen this issue before by chance? CC @walkingeyerobot

**Comment by sbc100** [comment]
acorn is indeed uses as part of the emscripten toolchain.      It should exist as part of the tar archive that the bazel toolchain (as well as the emsdk toolchain) use.
 
For example with emsdk I can see that it is included:

```
$ ./emsdk install 3.1.7
$ ls upstream/emscripten/node_modules/.bin/acorn 
upstream/emscripten/node_modules/.bin/acorn@
```

I'm not sure why it would not be found by the bazel toolchain since the same tar archive is used there.   @walkingeyerobot ?

**Comment by bkotsopoulossc** [comment]
I did some tests to see what change introduced the issue. 3.1.3 works fine, 3.1.4 has the issue. But if I use emsdk 3.1.4 with emscripten 3.1.3, that still has the issue. So it seems unrelated to emscripten changes, and has to do with some change in the emsdk bazel code between 3.1.3 and 3.1.4 (which looks to me like the nodejs upgrade is the culprit)

**Comment by bkotsopoulossc** [comment]
My hunch is this has something to do with bazel sandboxing mixing with something that changed in the new nodejs or npm version

**Comment by sbc100** [comment]
It looks like bazel and an npm install phase : https://github.com/emscripten-core/emsdk/blob/0ea8f8a8707070e9a7c83fbb4a3065683bcf1799/bazel/emscripten_deps.bzl#L73-L78

Does this appear to be working for you?  `acorn` is listed as dependency in the `emscirpten/package.json` file.

**Comment by bkotsopoulossc** [comment]
Yep but that was there before. The change was between 3.1.3 and 3.1.4. My hunch is this https://github.com/emscripten-core/emsdk/compare/3.1.3..3.1.4#diff-6d0d8b5e9320b3ad72eea2eabd547ed3cd2e2a39123bc7566a399168988747abR30 essentially

**Comment by mattsoulanille** [comment]
@bkotsopoulossc I haven't seen this before. Tested compiling [tfjs-backend-wasm](https://github.com/tensorflow/tfjs/tree/master/tfjs-backend-wasm) on Pop!_OS 21.10 x86_64 kernel 5.16.11 with the [3.1.4 tag](https://github.com/emscripten-core/emsdk/releases/tag/3.1.4) and [3.1.7 tag](https://github.com/emscripten-core/emsdk/releases/tag/3.1.7).

That being said, given the investigation you've done, I think you're right about that commit causing the issue. Something you could try to 100% verify this is setting node back to its original version by running rules_nodejs's [`node_repositories`](https://bazelbuild.github.io/rules_nodejs/install.html#installation-with-a-specific-version-of-nodejs-and-yarn) macro in your project's WORKSPACE before running emsdk's setup macros. This should [prevent node 16 from being used](https://github.com/emscripten-core/emsdk/blob/main/bazel/emscripten_deps.bzl#L27-L31). rules_nodejs version 4.4.1, which is [what emsdk currently uses](https://github.com/emscripten-core/emsdk/blob/main/bazel/deps.bzl#L10), has a [default node version of 14.17.5](https://github.com/bazelbuild/rules_nodejs/blob/7557d3c09edf1aa6c4abafc4616a3b008a19a547/nodejs/repositories.bzl#L11).

**Comment by bkotsopoulossc** [comment]
Yep tried that as well and verified it is true - I can upgrade successfully to 3.1.7 and force-downgrade nodejs back to what it was, and that works

**Comment by mattsoulanille** [comment]
I don't have the bandwidth right now to properly investigate why this is happening. For now, maybe we should downgrade Node back to its latest working version (i.e. remove [these lines](https://github.com/emscripten-core/emsdk/blob/main/bazel/emscripten_deps.bzl#L27-L31)) and just tell M1 Mac users that they need to manually specify a later version of Node that's compatible with Apple silicon. If their project uses Node or rules_nodejs, they've probably already done this anyway.

It's probably possible to automate this by using something like Bazel toolchains. If we create a [`node_toolchain`](https://bazelbuild.github.io/rules_nodejs/Core.html#node_toolchain) that uses node 16 and is compatible with only arm64 darwin, I think m1 macs will use this toolchain instead of the default one. I'm not sure if there's an easy way to do this, though, short of adding a step that downloads Node and pointing `node_toolchain` at it.

It's also possible that a later version of node works, but finding it without understanding the cause of the bug would probably be trial and error, which is not great.

Another alternative is we leave this as it is until a proper fix is found, since there seems to be a [workaround](https://github.com/emscripten-core/emsdk/issues/999#issuecomment-1067596690) for people who run into it.

Also, and this is likely a silly question, but I just thought I'd check. @bkotsopoulossc, you've tried `bazel clean --expunge`, right?

**Comment by bkotsopoulossc** [comment]
> It's also possible that a later version of node works,

I did a little bit of trial and error and found that 16.0.0 works (but had some other issues unrelated to bazel / emscripten).

> Another alternative is we leave this as it is until a proper fix is found

This is probably fine, as long as we make it easy for people to find the workaround. Maybe we leave this issue open

> you've tried bazel clean --expunge, right?

Yeah, before every build

**Comment by sbc100** [comment]
Closing since the bazel toolchain has evolved in many ways since this issue.  I don't think even use an npm install step anymore.
